### PR TITLE
match lowering: define a convenient struct

### PR DESCRIPTION
**Pull request by Nadrieril**
Small refactor PR: `bindings` and `ascriptions` always come together so I made a struct for them. I'll have one or two fields to add to it in a later PR as well.

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 1]
```diff
@@ -948,43 +955,38 @@ impl<'tcx, 'pat> FlatPat<'pat, 'tcx> {
...
 #[derive(Debug)]
 struct Candidate<'pat, 'tcx> {
-    /// [`Span`] of the original pattern that gave rise to this candidate.
-    span: Span,
-
-    /// Whether this `Candidate` has a guard.
-    has_guard: bool,
-
-    /// All of these must be satisfied...
+    /// For the candidate to match, &ll of these must be satisfied...
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:973]:
```suggestion
    /// For the candidate to match, all of these must be satisfied...
```

File: compiler/rustc_mir_build/src/build/matches/mod.rs [hunk 2]
```diff
@@ -1519,8 +1519,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
             // FIXME(or_patterns; matthewjasper) Try to be more aggressive here.
             can_merge &= subcandidate.subcandidates.is_empty()
-                && subcandidate.bindings.is_empty()
-                && subcandidate.ascriptions.is_empty();
+                && subcandidate.extra_data.bindings.is_empty()
+                && subcandidate.extra_data.ascriptions.is_empty();
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/mod.rs:1523]: Will the future information you add here be similar to these two lists and checked here? If so, maybe a single `is_empty` method on the struct could be useful

**Review by oli-obk — APPROVED** [review]

File: compiler/rustc_mir_build/src/build/matches/mod.rs (hunk 2 above)
Comment (Nadrieril) [reply compiler/rustc_mir_build/src/build/matches/mod.rs:1523]: Good shout :+1:

**Pull request merged**